R/functions-eager.R

```diff
@@ -1,25 +1,223 @@
-
-  elems <- list2(...)
-
-  if (length(elems) == 0L) {
-    abort("No arguments have been provided to concat.")
-  }
-
-  first <- elems[[1]]
-
-  if (length(elems) == 1L && (is_polars_df(first) || is_polars_series(first) || is_polars_lf(first))) {
-    return(first)
-  }
-
-  if (is_polars_df(first)) {
-    elems |>
-      lapply(\(x) x$`_df`) |>
-      concat_df() |>
-      wrap()
-  } else {
-    abort("todo")
-  }
+#' Combine multiple DataFrames, LazyFrames, Series, or Expr into a single object
+#'
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> DataFrames, LazyFrames,
+#' Series, or Expr to concatenate. All elements must have the same class.
+#' @param how Strategy to concatenate items. Must be one of:
+#' * `"vertical"`: applies multiple vstack operations;
+#' * `"vertical_relaxed"`: same as `"vertical"`, but additionally coerces
+#'   columns to their common supertype if they are mismatched (eg: Int32 to
+#'   Int64);
+#' * `"diagonal"`: finds a union between the column schemas and fills missing
+#'   column values with `null`;
+#' * `"diagonal_relaxed"`: same as `"diagonal"`, but additionally coerces
+#'   columns to their common supertype if they are mismatched (eg: Int32 to
+#'   Int64);
+#' * `"horizontal"`: stacks Series from DataFrames horizontally and fills with
+#'   `null` if the lengths don’t match;
+#' * `"align"`: Combines frames horizontally, auto-determining the common key
+#'   columns and aligning rows using the same logic as `align_frames`; this
+#'   behaviour is patterned after a full outer join, but does not handle
+#'   column-name collision. (If you need more control, you should use a
+#'   suitable join method instead).
+#'
+#' Series only support the `"vertical"` strategy.
+#' @param rechunk Make sure that the result data is in contiguous memory.
+#' @param parallel Only relevant for LazyFrames. This determines if the
+#' concatenated lazy computations may be executed in parallel.
+#'
+#' @return The same class (`polars_data_frame`, `polars_lazy_frame` or
+#' `polars_series`) as the input.
+#' @examples
+#' # default is 'vertical' strategy
+#' df1 <- pl$DataFrame(a = 1L, b = 3L)
+#' df2 <- pl$DataFrame(a = 2L, b = 4L)
+#' pl$concat(df1, df2)
+#'
+#' # 'a' is coerced to float64
+#' df1 <- pl$DataFrame(a = 1L, b = 3L)
+#' df2 <- pl$DataFrame(a = 2, b = 4L)
+#' pl$concat(df1, df2, how = "vertical_relaxed")
+#'
+#' df_h1 <- pl$DataFrame(l1 = 1:2, l2 = 3:4)
+#' df_h2 <- pl$DataFrame(r1 = 5:6, r2 = 7:8, r3 = 9:10)
+#' pl$concat(df_h1, df_h2, how = "horizontal")
+#'
+#' # use 'diagonal' strategy to fill empty column values with nulls
+#' df1 <- pl$DataFrame(a = 1L, b = 3L)
+#' df2 <- pl$DataFrame(a = 2L, c = 4L)
+#' pl$concat(df1, df2, how = "diagonal")
+#'
+#' df_a1 <- pl$DataFrame(id = 1:2, x = 3:4)
+#' df_a2 <- pl$DataFrame(id = 2:3, y = 5:6)
+#' df_a3 <- pl$DataFrame(id = c(1L, 3L), z = 7:8)
+#' pl$concat(df_a1, df_a2, df_a3, how = "align")
+pl__concat <- function(
+    ...,
+    how = "vertical",
```

Comment:
It's long but also the values that `how` can take depend on the input type (which is why the `arg_match0()` call comes late and isn't at the top of the function), so I don't think we should list them here.

Reply:
I understand that point of view, but I think it might be worth putting somethings for readability (since `arg_match0` can check for different default values for actual arguments).
For example, the `.names_repare` argument of the `as_tibble` function accepts input that is different from the default value (e.x. `"unique_quiet"`).
https://github.com/eitsupi/neo-r-polars/blob/0c54776fa5ba04c76382b978dfad5d4eecb841b4/R/dataframe-s3-tibble.R#L28

Reply:
I don't understand your argument. My point is that if we list all values of `how`, someone might think `"horizontal"` is accepted for all input types while it doesn't work with Series and therefore it is misleading . The code you link to only runs for `polars_data_frame` so there is no situation in which one of the values won't work because of the input type. 

>  For example, the .names_repare argument of the as_tibble function accepts input that is different from the default value (e.x. "unique_quiet").

This is unclear to me, do you have an example?

Reply:
> My point is that if we list all values of `how`, someone might think `"horizontal"` is accepted for all input types while it doesn't work with Series and therefore it is misleading

I understand that, but isn't that a problem because a friendly error message is displayed?

Reply:
Good point, I added the list of possible values in the function signature